File: src/main/cs/Problem0037.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Euler.test.cs;

namespace Euler.main.cs
{
    public class Problem0037
    {
        public Problem0037()
        {
            Stopwatch stopWatch = new Stopwatch();
            stopWatch.Start();
            long result = 0;

            // Solution:
//        result = solveSlow();
//        result = solveSlowButOptimized();

            //gen all primes < 1000000:
            List<int> primes = new List<int>(PrimeSieveButFast(1_000_000));

            List<string> firstPrimes = new List<string>{"2", "3", "5", "7"};
            for (int i = 0; i < 11;)
            {
                List<string> newPrimes = new List<string>();

                foreach (string prime in firstPrimes)
                {
                    for (int j = 0; j < 10; j++)
                    {
                        //generate primes:
                        string s = j + "" + prime;
                        if (primes.Contains(int.Parse(s)))
                        {
                            newPrimes.Add(s);
                        }

                        //check if valid:
                        for (int k = 1; k < s.Length; k++)
                        {
                            if (!primes.Contains(int.Parse(s.Substring(k))) ||
                                !primes.Contains(int.Parse(s.Substring(0, s.Length - k))))
                            {
                                goto end;
                            }
                        }

                        //when valid:
                        i++;
                        result += int.Parse(s);

                        end:
                        {
                        }
                    }
                }

                firstPrimes = newPrimes;
            }


            stopWatch.Stop();
            string elapsedTime = stopWatch.Elapsed.ToString();
            Console.WriteLine("Result:\t" + result + "\tTime:\t" +
                              (double.Parse(elapsedTime.Substring(elapsedTime.LastIndexOf(":") + 1, 2)) >= 1
                                  ? double.Parse(elapsedTime.Substring(elapsedTime.LastIndexOf(":") + 1)) + " s"
                                  : double.Parse(elapsedTime.Substring(elapsedTime.IndexOf(".") + 1)) / 10_000 +
                                    " ms"));
            if (Test.DoBenchmark)
            {
                Benchmark.AddTime(37, double.Parse(elapsedTime.Substring(elapsedTime.IndexOf(".") + 1)) / 10_000);
            }
        }

        private long solveSlowButOptimized()
        {
            long result = 0;

            int count = 0;
            List<int> primes = new List<int>(PrimeSieveButFast(1_000_000));
            foreach (int i in primes)
            {
                string teststring = i + "";

                //skip if <10 or Contains 5 (except leading)
                if (i < 10 ||
                    teststring.Substring(1).Contains("5"))
                {
                    continue;
                }

                //check if valid
                for (int j = 1; j < teststring.Length; j++)
                {
                    if (!primes.Contains(int.Parse(teststring.Substring(j))) ||
                        !primes.Contains(int.Parse(teststring.Substring(0, teststring.Length - j))))
                    {
                        goto end;
                    }
                }

                //when valid
                result += i;
                if (++count == 11)
                {
                    break;
                }

                end:
                {
                }
            }

            return result;
        }

        private long solveSlow()
        {
            long result = 0;
            for (int k = 2;; k++)
            {
                int count = 0;
                result = 0;
                List<int> primes = new List<int>(PrimeSieveButFast((int) Math.Pow(10.0, (double) k)));
                foreach (int i in primes)
                {
                    if (i < 10) continue;

                        string teststring = i + "";
                    for (int j = 1; j < teststring.Length; j++)
                    {
                        if (!primes.Contains(int.Parse(teststring.Substring(j))) ||
                            !primes.Contains(int.Parse(teststring.Substring(0, teststring.Length - j))))
                        {
                            goto middle;
                        }
                    }

                    result += i;
                    if (++count == 11)
                    {
                        goto end;
                    }

                    middle:
                    {
                    }
                }
            }

            end:
            {
            }
            return result;
        }

        private int[] PrimeSieveButFast(int range)
        {
            var bools = new bool[range + 1];
            Array.Fill(bools, true);
            var root = Math.Sqrt(range) + 0.5;
            for (var i = 3; i < root; i += 2)
                if (bools[i])
                    for (var j = i * i; j < range; j += i * 2)
                        bools[j] = false;

            var primes = new List<int> {2};
            for (var i = 3; i < range; i += 2)
                if (bools[i])
                    primes.Add(i);

            return primes.ToArray();
        }
    }
}

File: src/main/java/_00/Problem0037.java
package _00;

import java.util.ArrayList;
import java.util.Arrays;

public class Problem0037 {
    public Problem0037() {
        long startTime = System.nanoTime();
        long result = 0;

        // Solution:
//        result = solveSlow();
//        result = solveSlowButOptimized();

        //gen all primes < 1000000:
        ArrayList<Integer> primes = new ArrayList<Integer>(Arrays.asList(this.primeSieveButFast(1_000_000)));

        int count = 0;
        String[] firstPrimesArr = {"2", "3", "5", "7"};
        ArrayList<String> firstPrimes = new ArrayList<>(Arrays.asList(firstPrimesArr));
        for (int i = 0; i < 11; ) {
            ArrayList<String> newPrimes = new ArrayList<>();

            for (String prime : firstPrimes) {
                outer:
                for (int j = 0; j < 10; j++) {
                    //generate primes:
                    String s = j + "" + prime;
                    if (primes.contains(Integer.parseInt(s))) {
                        newPrimes.add(s);
                    }

                    //check if valid:
                    for (int k = 1; k < s.length(); k++) {
                        if (!primes.contains(Integer.parseInt(s.substring(k))) ||
                                !primes.contains(Integer.parseInt(s.substring(0, s.length() - k)))) {
                            continue outer;
                        }
                    }

                    //when valid:
                    i++;
                    result += Integer.parseInt(s);
                }
            }
            firstPrimes = newPrimes;
        }

        long timeToResolve = System.nanoTime() - startTime;
        System.out.println("Result:\t" + result + "\tTime:\t" + (((double) timeToResolve / 1_000_000) > 1000 ?
                (((double) timeToResolve / 1_000_000_000) + "s") :
                (((double) timeToResolve / 1_000_000) + "ms")));
    }

    private long solveSlowButOptimized() {
        long result = 0;

        int count = 0;
        ArrayList<Integer> primes = new ArrayList<Integer>(Arrays.asList(this.primeSieveButFast(1_000_000)));
        outer:
        for (Integer i : primes) {
            String testString = i + "";

            //skip if <10 or contains 5 (except leading)
            if (i < 10 ||
                    testString.substring(1).contains("5")) {
                continue;
            }

            //check if valid
            for (int j = 1; j < testString.length(); j++) {
                if (!primes.contains(Integer.parseInt(testString.substring(j))) ||
                        !primes.contains(Integer.parseInt(testString.substring(0, testString.length() - j)))) {
                    continue outer;
                }
            }

            //when valid
            result += i;
            if (++count == 11) {
                break;
            }
        }

        return result;
    }

    private long solveSlow() {
        long result = 0;
        outer:
        for (int k = 2; ; k++) {
            int count = 0;
            result = 0;
            ArrayList<Integer> primes = new ArrayList<Integer>(Arrays.asList(this.primeSieveButFast((int) Math.pow(10.0, k))));
            middle:
            for (Integer i : primes) {
                if (i < 10) {
                    continue;
                }
                String testString = i + "";
                for (int j = 1; j < testString.length(); j++) {
                    if (!primes.contains(Integer.parseInt(testString.substring(j))) ||
                            !primes.contains(Integer.parseInt(testString.substring(0, testString.length() - j)))) {
                        continue middle;
                    }
                }
                System.out.println(i);
                result += i;
                if (++count == 11) {
                    break outer;
                }
            }
        }
        return result;
    }

    private Integer[] primeSieveButFast(int range) {
        boolean[] bools = new boolean[range + 1];
        Arrays.fill(bools, true);
        double root = Math.sqrt(range) + 0.5;
        for (int i = 3; i < root; i += 2) {
            if (bools[i]) {
                for (int j = i * i; j < range; j += i * 2) {
                    bools[j] = false;
                }
            }
        }
        ArrayList<Integer> primes = new ArrayList<>();
        primes.add(2);
        for (int i = 3; i < range; i += 2) {
            if (bools[i]) {
                primes.add(i);
            }
        }
        return primes.toArray(new Integer[0]);
    }
}
